### PR TITLE
refreshing invisible widgets is a no-op

### PR DESCRIPTION
**Pull request by willmcgugan**
Fixes https://github.com/Textualize/textual/issues/5056
Reverts the work in https://github.com/Textualize/textual/pull/5048

Looking a little deeper regarding the issues with auto_refresh. The problem was when an automatic refresh occurs on a widget hidden by virtue of a parent being hidden, it still reset some internal data structures in the compositor.

The fix was to make refreshing widgets not on the screen a true no-up, which negates the need for https://github.com/Textualize/textual/pull/5048 -- which was fixing an issue this update prevents from occurring.

Thanks for the MRE @xavierog

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,12 +5,23 @@ All notable changes to this project will be documented in this file.
...
+
+- Added `DOMNode.is_on_screen` property https://github.com/Textualize/textual/pull/5063
+
 ## [0.81.0] - 2024-09-25
 
 ### Added
 
 - Added `x_axis` and `y_axis` parameters to `Widget.scroll_to_region` https://github.com/Textualize/textual/pull/5047
 - Added `Tree.move_cursor_to_line` https://github.com/Textualize/textual/pull/5052
+- Added `DOMNode.is_on_screen` property
```
Comment (xavierog) [inline CHANGELOG.md:24]: Duplicate line?

**Comment by darrenburns** [comment]
@willmcgugan It's not clear to me if this is ready for review or not

**Comment by willmcgugan** [comment]
Is ready for review

File: tests/snapshot_tests/test_snapshots.py [hunk 2]
```diff
@@ -2019,3 +2020,35 @@ def action_toggle_console(self) -> None:
 
     app = MRE()
     assert snap_compare(app, press=["space", "space", "z"])
+
+
+def test_updates_with_auto_refresh(snap_compare):
+    """Regression test for https://github.com/Textualize/textual/issues/5056"""
```
Comment (darrenburns) [inline tests/snapshot_tests/test_snapshots.py:2026]: This is hard to follow because it contains no context unless I click through to GitHub, and also contains some redundancy that might make sense in the MRE, but not in the test. If it fails, we should be able to validate the failure without following a paper-trail.

I think it's better to refine MREs a bit when translating them into tests.

```suggestion
    """Regression test for https://github.com/Textualize/textual/issues/5056
    
    After hiding and unhiding the RichLog, you should be able to see 1.5 fully rendered placeholder widgets.
    Prior to this fix, the bottom portion of the screen did not 
    refresh after the RichLog was hidden/unhidden while in the presence of the auto-refreshing ProgressBar widget.
    """
```

File: tests/snapshot_tests/test_snapshots.py [hunk 3]
```diff
@@ -2019,3 +2020,35 @@ def action_toggle_console(self) -> None:
...
+
+    class MRE(App):
+        BINDINGS = [
+            ("z", "toggle_widget('RichLog')", "Console"),
+            ("x", "toggle_widget('ProgressBar')", "Progress bar"),
```
Comment (darrenburns) [inline tests/snapshot_tests/test_snapshots.py:2031]: Appears unused.

```suggestion
```

File: tests/snapshot_tests/test_snapshots.py [hunk 4]
```diff
@@ -2019,3 +2020,35 @@ def action_toggle_console(self) -> None:
...
+        ]
+        CSS = """
+        Placeholder { height: 15; }
+        RichLog { border-top: dashed blue; height: 6; }
```
Comment (darrenburns) [inline tests/snapshot_tests/test_snapshots.py:2035]:
```suggestion
        RichLog { height: 6; }
```

File: tests/snapshot_tests/test_snapshots.py [hunk 5]
```diff
@@ -2019,3 +2020,35 @@ def action_toggle_console(self) -> None:
...
+        .hidden { display: none; }
+        """
+
+        def compose(self):
+            with VerticalScroll():
+                for i in range(10):
+                    yield Placeholder()
+            yield ProgressBar(classes="hidden")
+            yield RichLog(classes="hidden")
+            yield Footer()
```
Comment (darrenburns) [inline tests/snapshot_tests/test_snapshots.py:2045]:
```suggestion
```

**Review by darrenburns — CHANGES_REQUESTED** [review]

**Review by darrenburns — APPROVED** [review]
